// hammerhead/lib/errs.js
'use strict';

exports.__esModule = true;
var INJECTOR_RESOURCE_DECODING_FAILED = 'INJECTOR_RESOURCE_DECODING_FAILED';
exports.INJECTOR_RESOURCE_DECODING_FAILED = INJECTOR_RESOURCE_DECODING_FAILED;
var INJECTOR_RESOURCE_CHARSET_DECODING_FAILED = 'INJECTOR_RESOURCE_CHARSET_DECODING_FAILED';exports.INJECTOR_RESOURCE_CHARSET_DECODING_FAILED = INJECTOR_RESOURCE_CHARSET_DECODING_FAILED;
// TODO: remove
var INJECTOR_RESOURCE_ENCODING_FAILED = 'INJECTOR_RESOURCE_ENCODING_FAILED';

exports.INJECTOR_RESOURCE_ENCODING_FAILED = INJECTOR_RESOURCE_ENCODING_FAILED;
var PROXY_PORT_IS_ALREADY_IN_USE = 'PROXY_PORT_IS_ALREADY_IN_USE';exports.PROXY_PORT_IS_ALREADY_IN_USE = PROXY_PORT_IS_ALREADY_IN_USE;
// TODO: remove
var PROXY_ORIGIN_SERVER_CONNECTION_TERMINATED = 'PROXY_ORIGIN_SERVER_CONNECTION_TERMINATED';
exports.PROXY_ORIGIN_SERVER_CONNECTION_TERMINATED = PROXY_ORIGIN_SERVER_CONNECTION_TERMINATED;
var PROXY_CANT_RESOLVE_ORIGIN_URL = 'PROXY_CANT_RESOLVE_ORIGIN_URL';
exports.PROXY_CANT_RESOLVE_ORIGIN_URL = PROXY_CANT_RESOLVE_ORIGIN_URL;
var PROXY_XHR_REQUEST_SAME_ORIGIN_POLICY_VIOLATION = 'PROXY_XHR_REQUEST_SAME_ORIGIN_POLICY_VIOLATION';exports.PROXY_XHR_REQUEST_SAME_ORIGIN_POLICY_VIOLATION = PROXY_XHR_REQUEST_SAME_ORIGIN_POLICY_VIOLATION;
//TODO: remove
var PROXY_ORIGIN_SERVER_REQUEST_TIMEOUT = 'PROXY_ORIGIN_SERVER_REQUEST_TIMEOUT';
exports.PROXY_ORIGIN_SERVER_REQUEST_TIMEOUT = PROXY_ORIGIN_SERVER_REQUEST_TIMEOUT;
