// hammerhead/lib/router.js
'use strict';

var _classCallCheck = require('babel-runtime/helpers/class-call-check').default;

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default').default;

exports.__esModule = true;

var _sharedUrl_util = require('../shared/url_util');

var _utilsHash = require('./utils/hash');

var _utilsHash2 = _interopRequireDefault(_utilsHash);

var _utilsHttp = require('./utils/http');

// Const
var PARAM_RE = /^{(\S+)}$/;

// Static
function buildRouteParamsMap(routeMatch, paramNames) {
    return paramNames.reduce(function (params, paramName, i) {
        params[paramName] = routeMatch[i + 1];
        return params;
    }, {});
}

// Router

var Router = (function () {
    function Router() {
        _classCallCheck(this, Router);

        this.routes = {};
        this.routesWithParams = [];
    }

    Router.prototype._registerRoute = function _registerRoute(route, method, handler) {
        var tokens = route.split('/');
        var isRouteWithParams = tokens.some(function (token) {
            return PARAM_RE.test(token);
        });

        if (isRouteWithParams) this._registerRouteWithParams(tokens, method, handler);else {
            var isStatic = typeof handler !== 'function';

            if (isStatic) handler.etag = (0, _utilsHash2.default)(handler.content);

            this.routes['' + method + ' ' + route] = {
                handler: handler,
                isStatic: isStatic
            };
        }
    };

    Router.prototype._registerRouteWithParams = function _registerRouteWithParams(tokens, method, handler) {
        var paramNames = [];

        var reParts = tokens.map(function (token) {
            var paramMatch = token.match(PARAM_RE);

            if (paramMatch) {
                paramNames.push(paramMatch[1]);
                return '(\\S+?)';
            }

            return token;
        });
        this.routesWithParams.push({
            paramNames: paramNames,

            re: new RegExp('^' + method + ' ' + reParts.join('/') + '$'),
            handler: handler
        });
    };

    Router.prototype._route = function _route(req, res, serverInfo) {
        var routerQuery = '' + req.method + ' ' + (0, _sharedUrl_util.getPathname)(req.url);
        var route = this.routes[routerQuery];

        if (route) {
            if (route.isStatic) (0, _utilsHttp.respondStatic)(req, res, route.handler);else route.handler(req, res, serverInfo);

            return true;
        }

        for (var i = 0; i < this.routesWithParams.length; i++) {
            route = this.routesWithParams[i];

            var routeMatch = routerQuery.match(route.re);

            if (routeMatch) {
                var params = buildRouteParamsMap(routeMatch, route.paramNames);

                route.handler(req, res, serverInfo, params);
                return true;
            }
        }

        return false;
    };

    // API

    Router.prototype.GET = function GET(route, handler) {
        this._registerRoute(route, 'GET', handler);
    };

    Router.prototype.POST = function POST(route, handler) {
        this._registerRoute(route, 'POST', handler);
    };

    return Router;
})();

exports.default = Router;
module.exports = exports.default;
